// testcafe-hammerhead/lib/processing/script/node-builder.js
'use strict';

exports.__esModule = true;
exports.createStringLiteral = createStringLiteral;
exports.createTempVarIdentifier = createTempVarIdentifier;
exports.createAssignmentExprStmt = createAssignmentExprStmt;
exports.createBlockExprStmt = createBlockExprStmt;
exports.createVarDeclaration = createVarDeclaration;
exports.createProcessScriptMethCall = createProcessScriptMethCall;
exports.createLocationGetWrapper = createLocationGetWrapper;
exports.createLocationSetWrapper = createLocationSetWrapper;
exports.createPropertySetWrapper = createPropertySetWrapper;
exports.createMethCallWrapper = createMethCallWrapper;
exports.createPropertyGetWrapper = createPropertyGetWrapper;
exports.createComputedPropertyGetWrapper = createComputedPropertyGetWrapper;
exports.createComputedPropertySetWrapper = createComputedPropertySetWrapper;
exports.createGetEvalMethCall = createGetEvalMethCall;
exports.createGetPostMessageMethCall = createGetPostMessageMethCall;
exports.createExpandedConcatOperation = createExpandedConcatOperation;
exports.createHtmlProcessorWrapper = createHtmlProcessorWrapper;

var _esotope = require('./tools/esotope');

var _internalLiteral = require('./internal-literal');

var _internalLiteral2 = _interopRequireDefault(_internalLiteral);

var _instruction = require('./instruction');

var _instruction2 = _interopRequireDefault(_instruction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createStringLiteral(value) {
    return {
        type: _esotope.Syntax.Literal,
        value: value,
        raw: `"${value}"`
    };
} // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

function createTempVarIdentifier() {
    return {
        type: _esotope.Syntax.Identifier,
        name: _internalLiteral2.default.tempVar
    };
}

function createAssignmentExprStmt(left, right) {
    return {
        type: _esotope.Syntax.ExpressionStatement,

        expression: {
            type: _esotope.Syntax.AssignmentExpression,
            operator: '=',
            left: left,
            right: right
        }
    };
}

function createBlockExprStmt(children) {
    return {
        type: _esotope.Syntax.BlockStatement,
        body: children
    };
}

function createVarDeclaration(identifier, init) {
    return {
        type: _esotope.Syntax.VariableDeclaration,

        declarations: [{
            type: _esotope.Syntax.VariableDeclarator,
            id: identifier,
            init: init || null
        }],

        kind: 'var'
    };
}

function createProcessScriptMethCall(arg, isApply) {
    var ast = {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.processScript
        },

        arguments: [arg]
    };

    if (isApply) {
        ast.arguments.push({
            type: _esotope.Syntax.Literal,
            value: true,
            raw: 'true'
        });
    }

    return ast;
}

function createLocationGetWrapper() {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.getLocation
        },

        arguments: [{
            type: _esotope.Syntax.Identifier,
            name: 'location'
        }]
    };
}

function createLocationSetWrapper(value, wrapWithSequence) {
    var tempIdentifier = createTempVarIdentifier();
    var locationIdentifier = {
        type: _esotope.Syntax.Identifier,
        name: 'location'
    };

    var wrapper = {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.MemberExpression,
            computed: false,

            object: {
                type: _esotope.Syntax.FunctionExpression,
                id: null,
                params: [],
                defaults: [],

                body: {
                    type: _esotope.Syntax.BlockStatement,
                    body: [createVarDeclaration(tempIdentifier, value), {
                        type: _esotope.Syntax.ReturnStatement,

                        argument: {
                            type: _esotope.Syntax.LogicalExpression,
                            operator: '||',

                            left: {
                                type: _esotope.Syntax.CallExpression,

                                callee: {
                                    type: _esotope.Syntax.Identifier,
                                    name: _instruction2.default.setLocation
                                },

                                arguments: [locationIdentifier, tempIdentifier]
                            },

                            right: {
                                type: _esotope.Syntax.AssignmentExpression,
                                operator: '=',
                                left: locationIdentifier,
                                right: tempIdentifier
                            }
                        }
                    }]
                },

                rest: null,
                generator: false,
                expression: false
            },

            property: {
                type: _esotope.Syntax.Identifier,
                name: 'call'
            }
        },

        arguments: [{
            type: _esotope.Syntax.ThisExpression
        }]
    };

    if (wrapWithSequence) {
        wrapper = {
            type: _esotope.Syntax.SequenceExpression,

            expressions: [{
                type: 'Literal',
                value: 0,
                raw: '0'
            }, wrapper]
        };
    }

    return wrapper;
}

function createPropertySetWrapper(propertyName, obj, value) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.setProperty
        },

        arguments: [obj, createStringLiteral(propertyName), value]
    };
}

function createMethCallWrapper(owner, meth, args) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.callMethod
        },

        arguments: [owner, meth, {
            type: _esotope.Syntax.ArrayExpression,
            elements: args
        }]
    };
}

function createPropertyGetWrapper(propertyName, owner) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.getProperty
        },

        arguments: [owner, createStringLiteral(propertyName)]
    };
}

function createComputedPropertyGetWrapper(property, owner) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.getProperty
        },

        arguments: [owner, property]
    };
}

function createComputedPropertySetWrapper(property, owner, value) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.setProperty
        },

        arguments: [owner, property, value]
    };
}

function createGetEvalMethCall(node) {
    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.getEval
        },

        arguments: [node]
    };
}

function createGetPostMessageMethCall(node) {
    var parentObject = node.object;

    return {
        type: _esotope.Syntax.CallExpression,

        callee: {
            type: _esotope.Syntax.Identifier,
            name: _instruction2.default.getPostMessage
        },

        arguments: parentObject ? [parentObject] : [{
            type: _esotope.Syntax.Literal,
            value: null
        }, node]
    };
}

function createExpandedConcatOperation(left, right) {
    return {
        type: _esotope.Syntax.AssignmentExpression,
        operator: '=',
        left: left,

        right: {
            type: _esotope.Syntax.BinaryExpression,
            operator: '+',
            left: left,
            right: right
        }
    };
}

function createHtmlProcessorWrapper(node) {
    return {
        type: _esotope.Syntax.ExpressionStatement,

        expression: {
            type: _esotope.Syntax.CallExpression,

            callee: {
                type: _esotope.Syntax.MemberExpression,

                object: {
                    type: _esotope.Syntax.Identifier,
                    name: 'parent'
                },

                property: {
                    type: _esotope.Syntax.Identifier,
                    name: _instruction2.default.processHtml
                }
            },

            arguments: [{
                type: _esotope.Syntax.Identifier,
                name: 'window'
            }, node.expression]
        }
    };
}
