// testcafe-hammerhead/lib/utils/url.js
'use strict';

exports.__esModule = true;
exports.HTTPS_DEFAULT_PORT = exports.HTTP_DEFAULT_PORT = exports.SPECIAL_PAGES = exports.TRAILING_SLASH_RE = exports.REQUEST_DESCRIPTOR_VALUES_SEPARATOR = exports.HASH_RE = exports.SUPPORTED_PROTOCOL_RE = undefined;
exports.parseResourceType = parseResourceType;
exports.getResourceTypeString = getResourceTypeString;
exports.isSubDomain = isSubDomain;
exports.sameOriginCheck = sameOriginCheck;
exports.getURLString = getURLString;
exports.getProxyUrl = getProxyUrl;
exports.getDomain = getDomain;
exports.parseProxyUrl = parseProxyUrl;
exports.getPathname = getPathname;
exports.parseUrl = parseUrl;
exports.isSupportedProtocol = isSupportedProtocol;
exports.resolveUrlAsDest = resolveUrlAsDest;
exports.formatUrl = formatUrl;
exports.correctMultipleSlashes = correctMultipleSlashes;
exports.processSpecialChars = processSpecialChars;
exports.ensureTrailingSlash = ensureTrailingSlash;
exports.isSpecialPage = isSpecialPage;
exports.isRelativeUrl = isRelativeUrl;
exports.isValidUrl = isValidUrl;
exports.ensureOriginTrailingSlash = ensureOriginTrailingSlash;
exports.omitDefaultPort = omitDefaultPort;
exports.prepareUrl = prepareUrl;

var _stringTrim = require('./string-trim');

var _stringTrim2 = _interopRequireDefault(_stringTrim);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//Const
var PROTOCOL_RE = /^([\w-]+?:)(\/\/|[^\\/]|$)/; // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

var LEADING_SLASHES_RE = /^(\/\/)/;
var HOST_RE = /^(.*?)(\/|%|\?|;|#|$)/;
var PORT_RE = /:([0-9]*)$/;
var QUERY_AND_HASH_RE = /(\?.+|#[^#]*)$/;
var PATH_AFTER_HOST_RE = /^\/([^/]+?)\/([\S\s]+)$/;
var HTTP_RE = /^(?:https?):/;
var FILE_RE = /^file:/i;

var SUPPORTED_PROTOCOL_RE = exports.SUPPORTED_PROTOCOL_RE = /^(?:https?|file):/i;
var HASH_RE = exports.HASH_RE = /^#/;
var REQUEST_DESCRIPTOR_VALUES_SEPARATOR = exports.REQUEST_DESCRIPTOR_VALUES_SEPARATOR = '!';
var TRAILING_SLASH_RE = exports.TRAILING_SLASH_RE = /\/$/;
var SPECIAL_PAGES = exports.SPECIAL_PAGES = ['about:blank', 'about:error'];

var HTTP_DEFAULT_PORT = exports.HTTP_DEFAULT_PORT = '80';
var HTTPS_DEFAULT_PORT = exports.HTTPS_DEFAULT_PORT = '443';

function parseResourceType(resourceType) {
    if (!resourceType) {
        return {
            isIframe: false,
            isForm: false,
            isScript: false,
            isEventSource: false,
            isHtmlImport: false,
            isWebSocket: false
        };
    }

    return {
        isIframe: /i/.test(resourceType),
        isForm: /f/.test(resourceType),
        isScript: /s/.test(resourceType),
        isEventSource: /e/.test(resourceType),
        isHtmlImport: /h/.test(resourceType),
        isWebSocket: /w/.test(resourceType)
    };
}

function getResourceTypeString(resourceType) {
    resourceType = resourceType || {};

    if (!resourceType.isIframe && !resourceType.isForm && !resourceType.isScript && !resourceType.isEventSource && !resourceType.isHtmlImport && !resourceType.isWebSocket) return null;

    return [resourceType.isIframe ? 'i' : '', resourceType.isForm ? 'f' : '', resourceType.isScript ? 's' : '', resourceType.isEventSource ? 'e' : '', resourceType.isHtmlImport ? 'h' : '', resourceType.isWebSocket ? 'w' : ''].join('');
}

function isSubDomain(domain, subDomain) {
    domain = domain.replace(/^www./i, '');
    subDomain = subDomain.replace(/^www./i, '');

    if (domain === subDomain) return true;

    var index = subDomain.lastIndexOf(domain);

    return subDomain[index - 1] === '.' && subDomain.length === index + domain.length;
}

function sameOriginCheck(location, checkedUrl, rejectForSubdomains) {
    if (!checkedUrl) return true;

    var parsedLocation = parseUrl(location);
    var parsedCheckedUrl = parseUrl(checkedUrl);
    var parsedProxyLocation = parseProxyUrl(location);
    var parsedDestUrl = parsedProxyLocation ? parsedProxyLocation.destResourceInfo : parsedLocation;
    var isRelative = !parsedCheckedUrl.host;

    if (isRelative || parsedCheckedUrl.host === parsedLocation.host && parsedCheckedUrl.protocol === parsedLocation.protocol) return true;

    if (parsedDestUrl) {
        var portsEq = !parsedDestUrl.port && !parsedCheckedUrl.port || parsedDestUrl.port && parsedDestUrl.port.toString() === parsedCheckedUrl.port;

        if (parsedDestUrl.protocol === parsedCheckedUrl.protocol && portsEq) {
            if (parsedDestUrl.hostname === parsedCheckedUrl.hostname) return true;

            var isSubDomainHostname = isSubDomain(parsedDestUrl.hostname, parsedCheckedUrl.hostname) || isSubDomain(parsedCheckedUrl.hostname, parsedDestUrl.hostname);

            return !rejectForSubdomains && isSubDomainHostname;
        }
    }

    return false;
}

// NOTE: Convert the destination protocol and hostname to the lower case. (GH-1)
function convertHostToLowerCase(url) {
    var parsedUrl = parseUrl(url);
    var protocolHostSeparator = parsedUrl.protocol === 'about:' ? '' : '//';

    return (parsedUrl.protocol + protocolHostSeparator + parsedUrl.host).toLowerCase() + parsedUrl.partAfterHost;
}

function getURLString(url) {
    // TODO: fix it
    // eslint-disable-next-line no-undef
    if (url === null && /iPad|iPhone/i.test(window.navigator.userAgent)) return '';

    return String(url).replace(/\n|\t/g, '');
}

function getProxyUrl(url, opts) {
    var params = [opts.sessionId];

    if (opts.resourceType) params.push(opts.resourceType);

    if (opts.charset) params.push(opts.charset.toLowerCase());

    if (opts.reqOrigin) params.push(opts.reqOrigin);

    params = params.join(REQUEST_DESCRIPTOR_VALUES_SEPARATOR);

    var proxyProtocol = opts.proxyProtocol || 'http:';

    return proxyProtocol + '//' + opts.proxyHostname + ':' + opts.proxyPort + '/' + params + '/' + convertHostToLowerCase(url);
}

function getDomain(parsed) {
    return formatUrl({
        protocol: parsed.protocol,
        host: parsed.host,
        hostname: parsed.hostname,
        port: parsed.port
    });
}

function parseRequestDescriptor(desc) {
    var params = desc.split(REQUEST_DESCRIPTOR_VALUES_SEPARATOR);

    if (!params.length) return null;

    var sessionId = params[0];
    var resourceType = params[1] || null;
    var resourceData = params[2] || null;
    var parsedDesc = { sessionId, resourceType };

    if (resourceType && resourceData) {
        var parsedResourceType = parseResourceType(resourceType);

        if (parsedResourceType.isScript) parsedDesc.charset = resourceData;else if (parsedResourceType.isWebSocket) parsedDesc.reqOrigin = decodeURIComponent(resourceData);
    }

    return parsedDesc;
}

function parseProxyUrl(proxyUrl) {
    // TODO: Remove it.
    var parsedUrl = parseUrl(proxyUrl);

    if (!parsedUrl.partAfterHost) return null;

    var match = parsedUrl.partAfterHost.match(PATH_AFTER_HOST_RE);

    if (!match) return null;

    var parsedDesc = parseRequestDescriptor(match[1]);

    // NOTE: We should have, at least, the job uid and the owner token.
    if (!parsedDesc) return null;

    var destUrl = match[2];

    // Browser can redirect to a special page with hash (GH-1671)
    var destUrlWithoutHash = destUrl.replace(/#[\S\s]*$/, '');

    if (!isSpecialPage(destUrlWithoutHash) && !SUPPORTED_PROTOCOL_RE.test(destUrl)) return null;

    var destResourceInfo = !isSpecialPage(destUrlWithoutHash) ? parseUrl(match[2]) : {
        protocol: 'about:',
        host: '',
        hostname: '',
        port: '',
        partAfterHost: ''
    };

    return {
        destUrl: destUrl,
        destResourceInfo: destResourceInfo,
        partAfterHost: parsedUrl.partAfterHost,

        proxy: {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port
        },

        sessionId: parsedDesc.sessionId,
        resourceType: parsedDesc.resourceType,
        charset: parsedDesc.charset,
        reqOrigin: parsedDesc.reqOrigin
    };
}

function getPathname(path) {
    return path.replace(QUERY_AND_HASH_RE, '');
}

function parseUrl(url) {
    var parsed = {};

    url = processSpecialChars(url);

    if (!url) return parsed;

    url = (0, _stringTrim2.default)(url);

    // Protocol
    var hasImplicitProtocol = false;
    var remainder = url.replace(PROTOCOL_RE, function (str, protocol, strAfterProtocol) {
        parsed.protocol = protocol;
        return strAfterProtocol;
    }).replace(LEADING_SLASHES_RE, function () {
        hasImplicitProtocol = true;
        return '';
    });

    // NOTE: the URL is relative.
    if (!parsed.protocol && !hasImplicitProtocol) {
        parsed.partAfterHost = url;
        return parsed;
    }

    // Host
    parsed.partAfterHost = remainder.replace(HOST_RE, function (str, host, restPartSeparator) {
        parsed.host = host;
        parsed.port = '';
        return restPartSeparator;
    });

    parsed.hostname = parsed.host ? parsed.host.replace(PORT_RE, function (str, port) {
        parsed.port = port;
        return '';
    }) : '';

    return parsed;
}

function isSupportedProtocol(url) {
    url = (0, _stringTrim2.default)(url || '');

    var isHash = HASH_RE.test(url);

    if (isHash) return false;

    var protocol = url.match(PROTOCOL_RE);

    if (!protocol) return true;

    return SUPPORTED_PROTOCOL_RE.test(protocol[0]);
}

function resolveUrlAsDest(url, getProxyUrlMeth) {
    getProxyUrlMeth = getProxyUrlMeth || getProxyUrl;

    if (isSupportedProtocol(url)) {
        var proxyUrl = getProxyUrlMeth(url);
        var parsedProxyUrl = parseProxyUrl(proxyUrl);

        return parsedProxyUrl ? formatUrl(parsedProxyUrl.destResourceInfo) : url;
    }

    return url;
}

function formatUrl(parsedUrl) {
    // NOTE: the URL is relative.
    if (parsedUrl.protocol !== 'file:' && !parsedUrl.host && (!parsedUrl.hostname || !parsedUrl.port)) return parsedUrl.partAfterHost;

    var url = parsedUrl.protocol || '';

    url += '//';

    if (parsedUrl.username || parsedUrl.password) url += parsedUrl.username + ':' + parsedUrl.password + '@';

    if (parsedUrl.host) url += parsedUrl.host;else {
        url += parsedUrl.hostname;

        if (parsedUrl.port) url += ':' + parsedUrl.port;
    }

    if (parsedUrl.partAfterHost) url += parsedUrl.partAfterHost;

    return url;
}

function correctMultipleSlashes(url) {
    var pageProtocol = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

    // NOTE: Remove unnecessary slashes from the beginning of the url and after scheme.
    // For example:
    // "//////example.com" -> "//example.com" (scheme-less HTTP(S) URL)
    // "////home/testcafe/documents" -> "///home/testcafe/documents" (scheme-less unix file URL)
    // "http:///example.com" -> "http://example.com"
    //
    // And add missing slashes after the file scheme.
    // "file://C:/document.txt" -> "file:///C:/document.txt"
    if (url.match(FILE_RE) || pageProtocol.match(FILE_RE)) {
        return url.replace(/^(file:)?\/+(\/\/\/.*$)/i, '$1$2').replace(/^(file:)?\/*([A-Za-z]):/i, '$1///$2:');
    }

    return url.replace(/^(https?:)?\/+(\/\/.*$)/i, '$1$2');
}

function processSpecialChars(url) {
    return correctMultipleSlashes(getURLString(url));
}

function ensureTrailingSlash(srcUrl, processedUrl) {
    if (!isValidUrl(processedUrl)) return processedUrl;

    var srcUrlEndsWithTrailingSlash = TRAILING_SLASH_RE.test(srcUrl);
    var processedUrlEndsWithTrailingSlash = TRAILING_SLASH_RE.test(processedUrl);

    if (srcUrlEndsWithTrailingSlash && !processedUrlEndsWithTrailingSlash) processedUrl += '/';else if (!srcUrlEndsWithTrailingSlash && processedUrlEndsWithTrailingSlash) processedUrl = processedUrl.replace(TRAILING_SLASH_RE, '');

    return processedUrl;
}

function isSpecialPage(url) {
    return SPECIAL_PAGES.indexOf(url) !== -1;
}

function isRelativeUrl(url) {
    var parsedUrl = parseUrl(url);

    return parsedUrl.protocol !== 'file:' && !parsedUrl.host;
}

function isValidPort(port) {
    var parsedPort = parseInt(port, 10);

    return parsedPort > 0 && parsedPort <= 65535;
}

function isValidUrl(url) {
    var parsedUrl = parseUrl(url);

    return parsedUrl.protocol === 'file:' || parsedUrl.hostname && (!parsedUrl.port || isValidPort(parsedUrl.port));
}

function ensureOriginTrailingSlash(url) {
    // NOTE: If you request an url containing only port, host and protocol
    // then browser adds the trailing slash itself.
    var parsedUrl = parseUrl(url);

    if (!parsedUrl.partAfterHost && HTTP_RE.test(parsedUrl.protocol)) return url + '/';

    return url;
}

function omitDefaultPort(url) {
    // NOTE: If you request an url containing default port
    // then browser remove this one itself.
    var parsedUrl = parseUrl(url);

    var hasDefaultPort = parsedUrl.protocol === 'https:' && parsedUrl.port === HTTPS_DEFAULT_PORT || parsedUrl.protocol === 'http:' && parsedUrl.port === HTTP_DEFAULT_PORT;

    if (hasDefaultPort) {
        parsedUrl.host = parsedUrl.hostname;
        parsedUrl.port = '';

        return formatUrl(parsedUrl);
    }

    return url;
}

function prepareUrl(url) {
    url = omitDefaultPort(url);
    url = ensureOriginTrailingSlash(url);

    return url;
}
