// testcafe-browser-tools/lib/api/maximize.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _osFamily = require('os-family');

var _osFamily2 = _interopRequireDefault(_osFamily);

var _findWindow = require('./find-window');

var _findWindow2 = _interopRequireDefault(_findWindow);

var _utilsExec = require('../utils/exec');

var _binaries = require('../binaries');

var _binaries2 = _interopRequireDefault(_binaries);

function getBoundsFromString(boundsString) {
    return boundsString.split('\n');
}

function getWindowBounds(processId, windowId) {
    var boundsString;
    return _regeneratorRuntime.async(function getWindowBounds$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.getWindowBounds, [processId, windowId]));

            case 2:
                boundsString = context$1$0.sent;
                return context$1$0.abrupt('return', getBoundsFromString(boundsString));

            case 4:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}

function getWindowMaxBounds(processId, windowId) {
    var windowBounds, maxBoundsString;
    return _regeneratorRuntime.async(function getWindowMaxBounds$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(getWindowBounds(processId, windowId));

            case 2:
                windowBounds = context$1$0.sent;
                context$1$0.next = 5;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.getWindowMaxBounds, windowBounds));

            case 5:
                maxBoundsString = context$1$0.sent;
                return context$1$0.abrupt('return', getBoundsFromString(maxBoundsString));

            case 7:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}

function maximizeWindowMac(windowDescription) {
    var processId, windowId, windowBounds;
    return _regeneratorRuntime.async(function maximizeWindowMac$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                processId = windowDescription.processId;
                windowId = windowDescription.windowId;
                context$1$0.next = 4;
                return _regeneratorRuntime.awrap(getWindowMaxBounds(processId, windowId));

            case 4:
                windowBounds = context$1$0.sent;
                context$1$0.next = 7;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.setWindowBounds, [processId, windowId].concat(windowBounds)));

            case 7:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}
/**
 * Maximizes the specified browser window.
 * @function
 * @async
 * @name maximize
 * @param {string | object} windowDescriptor - The title of the web page opened in the window or a descriptor returned by findWindow.
 **/

exports.default = function callee$0$0(windowDescriptor) {
    var windowDescription, commandArguments;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                if (!(typeof windowDescriptor === 'string')) {
                    context$1$0.next = 6;
                    break;
                }

                context$1$0.next = 3;
                return _regeneratorRuntime.awrap(_findWindow2.default(windowDescriptor));

            case 3:
                context$1$0.t0 = context$1$0.sent;
                context$1$0.next = 7;
                break;

            case 6:
                context$1$0.t0 = windowDescriptor;

            case 7:
                windowDescription = context$1$0.t0;

                if (windowDescription) {
                    context$1$0.next = 10;
                    break;
                }

                return context$1$0.abrupt('return');

            case 10:
                commandArguments = void 0;

                if (!_osFamily2.default.win) {
                    context$1$0.next = 15;
                    break;
                }

                commandArguments = [windowDescription.hwnd];
                context$1$0.next = 26;
                break;

            case 15:
                if (!_osFamily2.default.linux) {
                    context$1$0.next = 19;
                    break;
                }

                commandArguments = [windowDescription.windowId];
                context$1$0.next = 26;
                break;

            case 19:
                if (!_osFamily2.default.mac) {
                    context$1$0.next = 25;
                    break;
                }

                context$1$0.next = 22;
                return _regeneratorRuntime.awrap(maximizeWindowMac(windowDescription));

            case 22:
                return context$1$0.abrupt('return');

            case 25:
                return context$1$0.abrupt('return');

            case 26:
                context$1$0.next = 28;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.maximize, commandArguments));

            case 28:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};

module.exports = exports.default;
