// testcafe-browser-tools/lib/utils/ensure-directory.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _gracefulFs = require('graceful-fs');

var _gracefulFs2 = _interopRequireDefault(_gracefulFs);

var _path = require('path');

var _mkdirp = require('mkdirp');

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _promisify = require('./promisify');

var _promisify2 = _interopRequireDefault(_promisify);

var _delay = require('./delay');

var _delay2 = _interopRequireDefault(_delay);

var stat = _promisify2.default(_gracefulFs2.default.stat);
var readdir = _promisify2.default(_gracefulFs2.default.readdir);
var mkdir = _promisify2.default(_mkdirp2.default);

var MAX_RETRY_COUNT = 10;
var RETRY_DELAY = 100;

function createDirectory(directoryPath) {
    return _regeneratorRuntime.async(function createDirectory$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.prev = 0;
                context$1$0.next = 3;
                return _regeneratorRuntime.awrap(mkdir(directoryPath));

            case 3:
                return context$1$0.abrupt('return', true);

            case 6:
                context$1$0.prev = 6;
                context$1$0.t0 = context$1$0['catch'](0);
                return context$1$0.abrupt('return', false);

            case 9:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this, [[0, 6]]);
}

function readDirectory(directoryPath) {
    return _regeneratorRuntime.async(function readDirectory$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.prev = 0;
                context$1$0.next = 3;
                return _regeneratorRuntime.awrap(readdir(directoryPath));

            case 3:
                return context$1$0.abrupt('return', true);

            case 6:
                context$1$0.prev = 6;
                context$1$0.t0 = context$1$0['catch'](0);
                return context$1$0.abrupt('return', false);

            case 9:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this, [[0, 6]]);
}

function ensureDirectory(directoryPath) {
    var fileEnsured, i;
    return _regeneratorRuntime.async(function ensureDirectory$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(createDirectory(directoryPath));

            case 2:
                context$1$0.t0 = context$1$0.sent;

                if (!context$1$0.t0) {
                    context$1$0.next = 7;
                    break;
                }

                context$1$0.next = 6;
                return _regeneratorRuntime.awrap(readDirectory(directoryPath));

            case 6:
                context$1$0.t0 = context$1$0.sent;

            case 7:
                fileEnsured = context$1$0.t0;
                i = 0;

            case 9:
                if (!(i < MAX_RETRY_COUNT && !fileEnsured)) {
                    context$1$0.next = 23;
                    break;
                }

                context$1$0.next = 12;
                return _regeneratorRuntime.awrap(_delay2.default(RETRY_DELAY));

            case 12:
                context$1$0.next = 14;
                return _regeneratorRuntime.awrap(createDirectory(directoryPath));

            case 14:
                context$1$0.t1 = context$1$0.sent;

                if (!context$1$0.t1) {
                    context$1$0.next = 19;
                    break;
                }

                context$1$0.next = 18;
                return _regeneratorRuntime.awrap(readDirectory(directoryPath));

            case 18:
                context$1$0.t1 = context$1$0.sent;

            case 19:
                fileEnsured = context$1$0.t1;

            case 20:
                i++;
                context$1$0.next = 9;
                break;

            case 23:
                return context$1$0.abrupt('return', fileEnsured);

            case 24:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}

exports.default = function callee$0$0(fileName) {
    var directoryPath, stats;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                directoryPath = _path.dirname(fileName);
                context$1$0.prev = 1;
                context$1$0.next = 4;
                return _regeneratorRuntime.awrap(stat(directoryPath));

            case 4:
                stats = context$1$0.sent;
                return context$1$0.abrupt('return', stats.isDirectory());

            case 8:
                context$1$0.prev = 8;
                context$1$0.t0 = context$1$0['catch'](1);

                if (!(context$1$0.t0.code === 'ENOENT')) {
                    context$1$0.next = 14;
                    break;
                }

                context$1$0.next = 13;
                return _regeneratorRuntime.awrap(ensureDirectory(directoryPath));

            case 13:
                return context$1$0.abrupt('return', context$1$0.sent);

            case 14:
                return context$1$0.abrupt('return', false);

            case 15:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this, [[1, 8]]);
};

module.exports = exports.default;
