// testcafe-hammerhead/lib/index.js
'use strict';

exports.__esModule = true;

var _proxy = require('./proxy');

var _proxy2 = _interopRequireDefault(_proxy);

var _session = require('./session');

var _session2 = _interopRequireDefault(_session);

var _responseMock = require('./request-pipeline/request-hooks/response-mock');

var _responseMock2 = _interopRequireDefault(_responseMock);

var _requestFilterRule = require('./request-pipeline/request-hooks/request-filter-rule');

var _requestFilterRule2 = _interopRequireDefault(_requestFilterRule);

var _script = require('./processing/script');

var _configureResponseEventOptions = require('./session/events/configure-response-event-options');

var _configureResponseEventOptions2 = _interopRequireDefault(_configureResponseEventOptions);

var _sameOriginCheckFailedStatusCode = require('./request-pipeline/xhr/same-origin-check-failed-status-code');

var _sameOriginCheckFailedStatusCode2 = _interopRequireDefault(_sameOriginCheckFailedStatusCode);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    Proxy: _proxy2.default,
    Session: _session2.default,
    processScript: _script.processScript,
    isScriptProcessed: _script.isScriptProcessed,
    ResponseMock: _responseMock2.default,
    RequestFilterRule: _requestFilterRule2.default,
    ConfigureResponseEventOptions: _configureResponseEventOptions2.default,
    SAME_ORIGIN_CHECK_FAILED_STATUS_CODE: _sameOriginCheckFailedStatusCode2.default
};
module.exports = exports['default'];
